pkg/cluster/workloadidentityresources.go

```diff
@@ -0,0 +1,114 @@
+// Licensed under the Apache License 2.0.
+
+import (
+	"fmt"
+
+	configv1 "github.com/openshift/api/config/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	azureFederatedTokenFileLocation = "/var/run/secrets/openshift/serviceaccount/token"
+
+	ccoSecretNamespace           = "openshift-cloud-credential-operator"
+	ccoSecretName                = "azure-credentials"
+	ccoSecretFilename            = "azure-ad-pod-identity-webhook-config.yaml"
+	authenticationConfigName     = "cluster"
+	authenticationConfigFilename = "cluster-authentication-02-config.yaml"
+)
+
+func (m *manager) generateWorkloadIdentityResources() (map[string]kruntime.Object, error) {
+	if !m.doc.OpenShiftCluster.UsesWorkloadIdentity() {
+		return nil, fmt.Errorf("generateWorkloadIdentityResources called for a CSP cluster")
+	}
+
+	resources := map[string]kruntime.Object{}
+	if platformWorkloadIdentitySecrets, err := m.generatePlatformWorkloadIdentitySecrets(); err != nil {
+		return nil, err
+	} else {
+		for _, secret := range platformWorkloadIdentitySecrets {
+			key := fmt.Sprintf("%s-%s-credentials.yaml", secret.ObjectMeta.Namespace, secret.ObjectMeta.Name)
+			resources[key] = secret
+		}
+	}
+
+	if cloudCredentialOperatorSecret, err := m.generateCloudCredentialOperatorSecret(); err != nil {
+		return nil, err
+	} else {
+		resources[ccoSecretFilename] = cloudCredentialOperatorSecret
+	}
+
+	if authenticationConfig, err := m.generateAuthenticationConfig(); err != nil {
+		return nil, err
+	} else {
+		resources[authenticationConfigFilename] = authenticationConfig
+	}
+
+	return resources, nil
+}
+
+func (m *manager) generatePlatformWorkloadIdentitySecrets() ([]*corev1.Secret, error) {
+	subscriptionId := m.subscriptionDoc.ID
+	tenantId := m.subscriptionDoc.Subscription.Properties.TenantID
+	region := m.doc.OpenShiftCluster.Location
+
+	roles := m.platformWorkloadIdentityRolesByVersion.GetPlatformWorkloadIdentityRolesByRoleName()
+
+	secrets := []*corev1.Secret{}
+	for _, identity := range m.doc.OpenShiftCluster.Properties.PlatformWorkloadIdentityProfile.PlatformWorkloadIdentities {
+		if role, ok := roles[identity.OperatorName]; ok {
+			secrets = append(secrets, &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: role.SecretLocation.Namespace,
+					Name:      role.SecretLocation.Name,
+				},
+				Type: corev1.SecretTypeOpaque,
+				StringData: map[string]string{
+					"azure_client_id":            identity.ClientID,
+					"azure_subscription_id":      subscriptionId,
+					"azure_tenant_id":            tenantId,
+					"azure_region":               region,
+					"azure_federated_token_file": azureFederatedTokenFileLocation,
+				},
+			})
+		}
+	}
+
+	return secrets, nil
+}
+
+func (m *manager) generateCloudCredentialOperatorSecret() (*corev1.Secret, error) {
+	tenantId := m.subscriptionDoc.Subscription.Properties.TenantID
+
+	return &corev1.Secret{
```

Comment:
Can we add and make sure the typeMeta is set when returning the object?
Maybe:-
```
TypeMeta: metav1.TypeMeta{
					Kind:       "Secret",
					APIVersion: corev1.SchemeGroupVersion.String(),
				},
```

Reply:
I'll do it for now, but I think there ought to be an elegant way to have our JSON/YAML serializer set the TypeMeta on the resources when writing them. This TypeMeta is not necessary when doing e.g. cluster updates, as K8S clients will automatically inject the TypeMeta on the resulting HTTP request bodies before sending them.

Reply:
Yup, can we have them generated based on isCreate flag somehow? In a way that the TypeMeta is only set for creation time manifests. When going through the installer code there are multiple places where it's just placed when returning the objects, eg:- https://github.com/openshift/installer/blob/release-4.14/pkg/asset/machines/baremetal/hosts.go#L39

Reply:
Apologies for un resolving it, wanted to make it visible.

Reply:
I don't think it's necessarily harmful to set it when doing updates, so I'll leave it alone. If I find that it causes problems later when implementing update, I can make it conditional then (or just remove TypeMeta before applying the resources).